features/steps/steps.py

```diff
@@ -413,3 +485,64 @@ def step_impl(context, entity, other_entities):
                 errors.append(instance_structure_error(inst, [i for i in nested_entities if i.is_a() in differences], 'nested by'))
     
     handle_errors(context, errors)
+
+def get_duplicate_values(values):
+    seen = set()
+    duplicates = [x for x in values if x in seen or seen.add(x)]
+    return duplicates
+
+def ifcopenshell_instance_type_to_string(v):
+    """
+        Converts ifcopenshell instance type to strings, if applicable
+    """
+    return do_try(lambda: v[0].is_a(), v)
+
+def empty_tuple_to_string(v):
+    """
+        Converts empty tuples type to strings, if applicable
+        To be used for meaningful error messages
+    """
+    return 'None' if isinstance(v, tuple) and not v else v
```

Comment:
Isn't using `'None'` here risky since it's a value that could also exist in the model. In which case we would get false positives or negatives?

Reply:
I also don't really understand why we need to convert things to string.

Reply:
Converting things to string is now used to compare entity types with each other. For example:
```
>>> f.by_type("IfcMapConversion")
[#24=IfcMapConversion(#11,#23,316131.64,5690966.11,$,1.,0.,$), #27=IfcMapConversion(#26,#23,316131.64,5690966.11,$,1.,0.,$)]
>>> f.by_type("IfcMapConversion")[0] == f.by_type("IfcMapConversion")[1]
False
>>> f.by_type("IfcMapConversion")[0].is_a() == f.by_type("IfcMapConversion")[1].is_a() 
True #'IfcMapConversion" == "IfcMapConversion"
```

Or is there a better way to compare entities?

Also, sometimes IfcGeometricRepresentationContext gives back no value (or, to be more precise, an empty tuple). In that case, it would be more informative to return 'None' instead of '()'. But again, maybe this function is too specific for this rule and therefore should be modified to be more reusable. 

I could also leave the work for error messaging out and leave it for the non-rule sprint. For example, since relating (#23IfcMapConversion... and the value IfcMapConversion) are so closely related, I left the entity in which the value was found out of the error message. In GEM003 for example, stating at which entity value 'Axis' or 'Body' for RepresentationIdentifier was found is very informative, but in this rule less (I think ..). What do you think?

Reply:
I also realized we did something comparable with the project for Schöck, by using hashing. We can also compare instances that we and be as specific as we want. 

It also doesn't take into inheritance now I see now. I'll change it that it will take that into consideration, so it is in line with the approach we agreed earlier on (entity.is_a(i)). And take a note for a next meeting, I can imagine comparing entities will come back in the future, and I don't know if simple is good or simplistic..
```
>>> mapconv[0].is_a('IfcRigidOperation')
False
>>> mapconv[0].is_a('IfcMapConversion')
True
>>> mapconv[0].is_a('IfcCoordinateOperation')
True
```

Reply:
There is an option to consider inheritance now that we can reuse in future rules https://github.com/buildingSMART/ifc-gherkin-rules/pull/35/commits/7237bcaf90edf74e82548037b7ed2220903dbbc4

Reply:
Hold on a sec. When we say "The values must be identical" it's not just that they should both be *of type IfcMapConversion*, they should point to the exact same instance or have exact identical values. You can use `get_info(recursive=True, include_identifiers=False)` to get a native python dict that you can use for "deep" equality testing.